### ISSUE TITLE
support long etcd compaction duration at k3s startup

### ISSUE THREAD

**Issue by gberche-orange** [k3s-io/k3s#11241]
**Is your feature request related to a problem? Please describe.**

There are conditions (detailed in *Additional context*) in which the database grows beyond the 30s timeout enforced by k3s both for scheduled compactions and k3s startup. As a result, for a k3s cluster with significant etcd write traffic, the etcd database would quickly grow and reach the configured quota, preventing any more k8s api writes.

Trying to recover the situation by increasing the etcd quota and triggering compaction by k3s restart isn't sufficient, as k3s still fails after a 30s compaction timeout.

<details>

<summary>logs of looping defragmentation failures at k3s startup</summary>

```
{"level":"warn","ts":"2024-10-31T10:19:00.822892Z","logger":"etcd-client","caller":"v3@v3.5.13-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00117c3c0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
time="2024-10-31T10:19:00Z" level=error msg="Failed to check local etcd status for learner management: context deadline exceeded"
{"level":"info","ts":"2024-10-31T10:19:08.735807Z","caller":"backend/backend.go:549","msg":"finished defragmenting directory","path":"/var/vcap/store/k3s-server/server/db/etcd/member/snap/db","current-db-size-bytes-diff":0,"current-db-size-bytes":4867026944,"current-db-size":"4.9 GB","current-db-size-in-use-bytes-diff":0,"current-db-size-in-use-bytes":4867006464,"current-db-size-in-use":"4.9 GB","took":"42.268660727s"}
{"level":"info","ts":"2024-10-31T10:19:08.735941Z","caller":"v3rpc/maintenance.go:96","msg":"finished defragment"}
{"level":"warn","ts":"2024-10-31T10:19:08.736001Z","caller":"v3rpc/interceptor.go:197","msg":"request stats","start time":"2024-10-31T10:18:26.467053Z","time spent":"42.268936159s","remote":"127.0.0.1:57122","response type":"/etcdserverpb.Maintenance/Defragment","request count":-1,"request size":-1,"response count":-1,"response size":-1,"request content":""}
{"level":"warn","ts":"2024-10-31T10:19:08.736099Z","caller":"v3rpc/interceptor.go:197","msg":"request stats","start time":"2024-10-31T10:18:30.825006Z","time spent":"37.911063176s","remote":"127.0.0.1:47338","response type":"/etcdserverpb.Maintenance/Status","request count":-1,"request size":-1,"response count":-1,"response size":-1,"request content":""}
{"level":"warn","ts":"2024-10-31T10:19:08.737863Z","caller":"v3rpc/interceptor.go:197","msg":"request stats","start time":"2024-10-31T10:18:45.826224Z","time spent":"22.911629949s","remote":"127.0.0.1:58580","response type":"/etcdserverpb.Maintenance/Status","request count":-1,"request size":-1,"response count":-1,"response size":-1,"request content":""}
{"level":"warn","ts":"2024-10-31T10:19:08.737881Z","caller":"v3rpc/interceptor.go:197","msg":"request stats","start time":"2024-10-31T10:18:54.227529Z","time spent":"14.510343874s","remote":"127.0.0.1:60794","response type":"/etcdserverpb.Maintenance/Status","request count":-1,"request size":-1,"response count":-1,"response size":-1,"request content":""}
{"level":"warn","ts":"2024-10-31T10:19:08.737903Z","caller":"v3rpc/interceptor.go:197","msg":"request stats","start time":"2024-10-31T10:19:00.828874Z","time spent":"7.909022843s","remote":"127.0.0.1:54000","response type":"/etcdserverpb.Maintenance/Status","request count":-1,"request size":-1,"response count":-1,"response size":-1,"request content":""}
time="2024-10-31T10:19:08Z" level=info msg="Defragmenting etcd database"
{"level":"info","ts":"2024-10-31T10:19:08.7433Z","caller":"v3rpc/maintenance.go:90","msg":"starting defragment"}
{"level":"info","ts":"2024-10-31T10:19:08.777761Z","caller":"snap/snapshotter.go:293","msg":"found orphaned .snap.db file; deleting","path":"0000000041760ad9.snap.db"}
{"level":"info","ts":"2024-10-31T10:19:08.783933Z","caller":"backend/backend.go:497","msg":"defragmenting","path":"/var/vcap/store/k3s-server/server/db/etcd/member/snap/db","current-db-size-bytes":4867026944,"current-db-size":"4.9 GB","current-db-size-in-use-bytes":4867010560,"current-db-size-in-use":"4.9 GB"}
{"level":"info","ts":"2024-10-31T10:19:09.565023Z","caller":"etcdserver/server.go:2464","msg":"saved snapshot","snapshot-index":1098265088}
{"level":"info","ts":"2024-10-31T10:19:09.565407Z","caller":"etcdserver/server.go:2494","msg":"compacted Raft logs","compact-index":1098260088}
time="2024-10-31T10:19:09Z" level=info msg="Waiting for etcd server to become available"
time="2024-10-31T10:19:09Z" level=info msg="Waiting for API server to become available"
{"level":"warn","ts":"2024-10-31T10:19:15.823589Z","logger":"etcd-client","caller":"v3@v3.5.13-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00117c3c0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
time="2024-10-31T10:19:15Z" level=error msg="Failed to get recorded learner progress from etcd: context deadline exceeded"
{"level":"info","ts":"2024-10-31T10:19:17.98241Z","caller":"etcdserver/server.go:1443","msg":"triggering snapshot","local-member-id":"ec15cc28e3bfecca","local-member-applied-index":1098275109,"local-member-snapshot-index":1098265088,"local-member-snapshot-count":10000}
{"level":"warn","ts":"2024-10-31T10:19:24.223795Z","logger":"etcd-client","caller":"v3@v3.5.13-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00117c3c0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
time="2024-10-31T10:19:24Z" level=info msg="Failed to test data store connection: failed to defragment etcd database: context deadline exceeded"
```

</details>

The manual workaround we applied to recover this situation in a k3s cluster with multiple servers (and no dedicated etcd nodes) was:
- stop the whole k3s cluster
- choose one of the healthy k3s etcd server, if possible the one calling the `--cluster-init` arg at startup 
- try offline defrag using `etcdutl` cli (see https://docs.k3s.io/advanced#using-etcdctl): succeeds after 2m14s but still missing compaction, so the db is still too large (5.1GB)

<details>

<summary>traces of `etcdutl defrag` command</summary>

```
server/52168ef9-175e-4f17-9b6c-ecbc23193746:/var/vcap/store/k3s-server/server/db/snapshots# etcdutl defrag --data-dir /var/vcap/store/k3s-server/server/db/etcd
waiting for etcd to close and release its lock on "/var/vcap/store/k3s-server/server/db/etcd/member/snap/db". To defrag a running etcd instance, omit --data-dir.
2024-10-31T13:15:13Z    info    backend/backend.go:497  defragmenting   {"path": "/var/vcap/store/k3s-server/server/db/etcd/member/snap/db", "current-db-size-bytes": 5125394432, "current-db-size": "5.1 GB", "current-db-size-in-use-bytes": 5124485120, "current-db-size-in-use": "5.1 GB"}
2024-10-31T13:17:27Z    info    backend/backend.go:549  finished defragmenting directory        {"path": "/var/vcap/store/k3s-server/server/db/etcd/member/snap/db", "current-db-size-bytes-diff": -1429504, "current-db-size-bytes": 5123964928, "current-db-size": "5.1 GB", "current-db-size-in-use-bytes-diff": -552960, "current-db-size-in-use-bytes": 5123932160, "current-db-size-in-use": "5.1 GB", "took": "2m14.086870095s"}

server/52168ef9-175e-4f17-9b6c-ecbc23193746:/var/vcap/store/k3s-server/server/db/snapshots# time etcdutl defrag --data-dir /var/vcap/store/k3s-server/server/db/etcd
waiting for etcd to close and release its lock on "/var/vcap/store/k3s-server/server/db/etcd/member/snap/db". To defrag a running etcd instance, omit --data-dir.
2024-10-31T13:18:48Z    info    backend/backend.go:497  defragmenting   {"path": "/var/vcap/store/k3s-server/server/db/etcd/member/snap/db", "current-db-size-bytes": 5123964928, "current-db-size": "5.1 GB", "current-db-size-in-use-bytes": 5123932160, "current-db-size-in-use": "5.1 GB"}
2024-10-31T13:19:15Z    info    backend/backend.go:549  finished defragmenting directory        {"path": "/var/vcap/store/k3s-server/server/db/etcd/member/snap/db", "current-db-size-bytes-diff": 0, "current-db-size-bytes": 5123964928, "current-db-size": "5.1 GB", "current-db-size-in-use-bytes-diff": 0, "current-db-size-in-use-bytes": 5123932160, "current-db-size-in-use": "5.1 GB", "took": "26.429924344s"}

real    0m28.493s
user    0m8.738s
sys     0m13.154s
```

 
</details>

- start a local etcd server with `--force-new-cluster` flag to ignore other members found in store: `etcd --data-dir /var/vcap/store/k3s-server/server/db/etcd --force-new-cluster` 

- use 3rd party etcd-defrag to trigger compaction, available at https://github.com/ahrtr/etcd-defrag

<details>

<summary>detailed used of etcd-defrag, succeeds in 4s</summary>

```
server/52168ef9-175e-4f17-9b6c-ecbc23193746:/var/vcap/store/k3s-server/server/db/etcd# etcdctl --dial-timeout=20s --command-timeout=50s endpoint status --cluster --write-out=table
+-----------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
|       ENDPOINT        |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
+-----------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
| http://localhost:2379 | 550b2c4d3542b3aa |  3.5.13 |  5.1 GB |      true |      false |       731 | 1098985026 |         1098985026 |        |
+-----------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
server/52168ef9-175e-4f17-9b6c-ecbc23193746:/var/vcap/store/k3s-server/server/db/etcd# etcd-defrag -h
A simple command line tool for etcd defragmentation

Usage:
  etcd-defrag [flags]

Flags:
      --cacert string                  verify certificates of TLS-enabled secure servers using this CA bundle
      --cert string                    identify secure client using this TLS certificate file
      --cluster                        use all endpoints from the cluster member list
      --command-timeout duration       command timeout (excluding dial timeout) (default 30s)
      --compaction                     whether execute compaction before the defragmentation (defaults to true) (default true)
      --continue-on-error              whether continue to defragment next endpoint if current one fails (default true)
      --defrag-rule string             defragmentation rule (etcd-defrag will run defragmentation if the rule is empty or it is evaluated to true)
      --dial-timeout duration          dial timeout for client connections (default 2s)
  -d, --discovery-srv string           domain name to query for SRV records describing cluster endpoints
      --discovery-srv-name string      service name to query when using DNS discovery
      --dry-run                        evaluate whether or not endpoints require defragmentation, but don't actually perform it
      --endpoints strings              comma separated etcd endpoints (default [127.0.0.1:2379])
      --etcd-storage-quota-bytes int   etcd storage quota in bytes (the value passed to etcd instance by flag --quota-backend-bytes) (default 2147483648)
      --exclude-localhost              whether to exclude localhost endpoints
  -h, --help                           help for etcd-defrag
      --insecure-discovery             accept insecure SRV records describing cluster endpoints (default true)
      --insecure-skip-tls-verify       skip server certificate verification (CAUTION: this option should be enabled only for testing purposes)
      --insecure-transport             disable transport security for client connections (default true)
      --keepalive-time duration        keepalive time for client connections (default 2s)
      --keepalive-timeout duration     keepalive timeout for client connections (default 6s)
      --key string                     identify secure client using this TLS key file
      --password string                password for authentication (if this option is used, --user option shouldn't include password)
      --user string                    username[:password] for authentication (prompt if password is not supplied)
      --version                        print the version and exit

server/52168ef9-175e-4f17-9b6c-ecbc23193746:/var/vcap/store/k3s-server/server/db/etcd# etcd-defrag 
Validating configuration.
No defragmentation rule provided
Performing health check.
endpoint: http://localhost:2379, health: true, took: 6.875522ms, error: 
Getting members status
endpoint: 127.0.0.1:2379, dbSize: 5124136960, dbSizeInUse: 5124120576, memberId: 550b2c4d3542b3aa, leader: 550b2c4d3542b3aa, revision: 700878502, term: 731, index: 1098985027
Running compaction until revision: 700878502 ... successful
1 endpoint(s) need to be defragmented: [127.0.0.1:2379]
[Before defragmentation] endpoint: 127.0.0.1:2379, dbSize: 5125455872, dbSizeInUse: 229502976, memberId: 550b2c4d3542b3aa, leader: 550b2c4d3542b3aa, revision: 700878502, term: 731, index: 1098985028
Defragmenting endpoint "127.0.0.1:2379"
Finished defragmenting etcd endpoint "127.0.0.1:2379". took 3.54581835s
[Post defragmentation] endpoint: 127.0.0.1:2379, dbSize: 233582592, dbSizeInUse: 233512960, memberId: 550b2c4d3542b3aa, leader: 550b2c4d3542b3aa, revision: 700878502, term: 731, index: 1098985028
The defragmentation is successful.


server/52168ef9-175e-4f17-9b6c-ecbc23193746:/var/vcap/store/k3s-server/server/db/etcd# etcdctl --dial-timeout=20s --command-timeout=50s endpoint status --cluster --write-out=table
+-----------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
|       ENDPOINT        |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
+-----------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
| http://localhost:2379 | 550b2c4d3542b3aa |  3.5.13 |  234 MB |      true |      false |       731 | 1098985028 |         1098985028 |        |
+-----------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+

```
 
 
</details>

- stop local etcd server
- start k3s server, possibly with `--cluster-reset` flag as mentionned into https://docs.k3s.io/cli/etcd-snapshot?_highlight=cluster&_highlight=reset&etcdsnap=High+Availability
- observe that the scheduled etcd compactions now succeeds

<details>

<summary>successfull scheduled etcd compaction traces every 5 mins</summary>

```
{"level":"info","ts":"2024-10-31T15:47:38.995335Z","caller":"mvcc/kvstore_compaction.go:68","msg":"finished scheduled compaction","compact-revision":701073902,"took":"3.815544534s","hash":1594712287,"current-db-size-bytes":1203703808,"current-db-size":"1.2 GB","current-db-size-in-use-bytes":349925376,"current-db-size-in-use":"350 MB"}
[...]
{"level":"info","ts":"2024-10-31T15:52:37.838455Z","caller":"mvcc/kvstore_compaction.go:68","msg":"finished scheduled compaction","compact-revision":701093922,"took":"2.660997733s","hash":904942854,"current-db-size-bytes":1203703808,"current-db-size":"1.2 GB","current-db-size-in-use-bytes":317558784,"current-db-size-in-use":"318 MB"}
```
 
 
</details>

- remove etcd db from other servers (rm or mv the `${data-dir}/server/db/` directory)
- start other servers one by one, and observe that they synchronize to the etcd leader

**Describe the solution you'd like**

Ability to transiently specify the value of the `testTimeout` at k3s startup with a dedicated flag. 

Currently, this duration is hard coded to 30s, see https://github.com/k3s-io/k3s/blob/b0bf7e765c2f748cbfed2dc308526243c037660f/pkg/etcd/etcd.go#L58 and seems used both for initial compaction at startup and scheduled compactions.

[screenshot testTimeout usage](https://github.com/user-attachments/assets/5cc5c785-82f8-410d-a1c4-ab72ddb6181f)
[screenshot Test call hiearchy at startup](https://github.com/user-attachments/assets/7bcc91eb-6f06-477b-93c4-3bf3fb85ad5e)

See related previous discussions at  https://github.com/k3s-io/k3s/issues/5881#issuecomment-1322550483

**Describe alternatives you've considered**

Introduce distinct longer hardcoded timeout for compaction at startup.

**Additional context**

Scenario to that led our k3s cluster to have its etcd db grow beyond the 30s timeout for compaction:

A configuration error in the kube scheduler (passed through the `--kube-scheduler-arg=config`) propagates to the whole k3s servers which triggers their exit, few seconds after serving k3s api traffic. Our automation restarts the k3s servers immediately. This partial failed start loop repeats on our 3 k3s servers, which still accept k8s api write traffic for a while without applying defrag nor compaction.

After 2 hours, our etcd 5GB quota is reached and prevents new k8s api writes. 

![image](https://github.com/user-attachments/assets/6723fa35-dcaa-4df6-b56d-3de660dcd5df)
![image](https://github.com/user-attachments/assets/996ab6f7-ffe3-4461-bd69-ab5e8dac35a4)

**Comment by brandond** [comment]
* Duplicate of https://github.com/k3s-io/k3s/issues/11122

**Comment by gberche-orange** [comment]
@brandond I see that https://github.com/k3s-io/k3s/issues/11122 has finally been marked as "could not reproduce, won't fix" in https://github.com/k3s-io/k3s/issues/11122#issuecomment-2470303374

Are you then suggesting k3s users experiencing online compaction failures on their infrastructure to perform offline compactions after stopping k3s using native etcd tooling ?

Any chance for you to reconsider introducing a higher or configurable compaction timeout for users where the problem reproduces on their infrastructure ?

**Comment by brandond** [comment]
See https://github.com/k3s-io/k3s/issues/11122#issuecomment-2433446640 and https://github.com/k3s-io/k3s/pull/11164

Note that it was NOT closed as won't fix, you are making an incorrect assumption. QA simply noted that they were unable to test the fix because they could not create the poor disk performance required to exceed the timeout.

**Comment by gberche-orange** [comment]
Thanks for your response, and sorry for the misunderstanding and the associated noise.

The associated commit https://github.com/k3s-io/k3s/commit/095e34d816e672cfc579a93e7dacab5bfc0144b0 from #11164 is merged and available in the following releases (did not find reference to #11164 while browsing https://github.com/k3s-io/k3s/releases or individual release such as https://github.com/k3s-io/k3s/releases/tag/v1.32.3-rc4+k3s1

![Image](https://github.com/user-attachments/assets/773b3382-c17d-4602-a005-99170b549ff1)
![Image](https://github.com/user-attachments/assets/3f06849d-0fe1-48e2-a6e1-ab13efd1f97b)

**Comment by brandond** [comment]
Scroll down to the bottom of that PR, where you will see a number of bulk backport PRs for the 2024-11 cycle linked, one per release branch. Shouldn't be hard to find.

**Comment by gberche-orange** [comment]
Thanks ! 

https://github.com/k3s-io/k3s/pull/11164
![Image](https://github.com/user-attachments/assets/1c32e122-dcd1-4ded-b5f0-964f762a6124)

therefore 1.30 includes the fix starting 1.30.7
https://github.com/k3s-io/k3s/pull/11262
![Image](https://github.com/user-attachments/assets/1ea7565e-e9e3-4d59-ab74-d3b9078cd344)
https://github.com/k3s-io/k3s/commit/185400d479031a684d33919822bac80b181650d9
![Image](https://github.com/user-attachments/assets/5620a25f-e08d-4b51-84fa-d3b21336e241)

https://github.com/k3s-io/k3s/pull/11261 1.31 includes the fix starting 1.31.3
https://github.com/k3s-io/k3s/commit/4e86699dfd6d5043e118906fbc85a744db260ce5

![Image](https://github.com/user-attachments/assets/8816e58f-68bb-4053-9d1b-9004248a8750)
